orttraining/orttraining/test/python/orttraining_test_ortmodule_fallback.py

```diff
@@ -554,14 +554,57 @@ def test_ortmodule_fallback_warn_message(is_training, persist_fallback):
     # Use data in different device for testing
     inputs = torch.randn(N, D_in, device=data_device)
 
-    for _ in range(3):
+    for i in range(3):
         with pytest.raises(RuntimeError), pytest.warns(UserWarning) as warning_record:
             ort_model(inputs)
-        assert "Fallback to PyTorch due to exception" in str(warning_record[0].message.args[0])
+
+        # Only log the warn message once or when retries is enabled
+        # Otherwise the warn message will be skipped
+        if i == 0 or not persist_fallback:
+            assert "Fallback to PyTorch due to exception" in str(warning_record[0].message.args[0])
+        else:
+            assert warning_record.list == []
 
```

Comment:
I was thinking something like this

```suggestion
    with pytest.raises(RuntimeError), pytest.warns(UserWarning) as warning_record:
        ort_model(inputs)
    assert "Fallback to PyTorch due to exception" in str(warning_record[0].message.args[0])
    
    with pytest.raises(RuntimeError), pytest.warns(UserWarning) as warning_record:
        ort_model(inputs)
    if not persist_fallback:
        assert "Fallback to PyTorch due to exception" in str(warning_record[0].message.args[0])
    else:
        # Only log the warn message once or when retries is enabled
        # Otherwise the warn message will be skipped
        assert warning_record.list == []

```

so that there is no for loops and index checking. I do see we are duplicating long asserts as a result. Feel free to disregard this if you feel it doesn't improve redability.

Reply:
All I see, thank you for your explanation.  It totally makes sense. However, above code the first `assert` `warning_record[0]` will have `IndexError: list index out of range` if we do not log the warn message.  Based on your explanation, how about below code changes? 
```
    for i in range(3):
        with pytest.raises(RuntimeError), pytest.warns(UserWarning) as warning_record:
            ort_model(inputs)
        # For `retries`, the warn message will always be logged
        if not persist_fallback:
            assert assert "Fallback to PyTorch due to exception" in str(warning_record[0].message.args[0])
            continue
    
        # If `retries` is not enabled, only log the warn message once
        if i == 0:
            assert "Fallback to PyTorch due to exception" in str(warning_record[0].message.args[0])
        else:
            assert warning_record.list == []
```

Reply:
looks good to me. thanks!